### DISCUSSION TITLE
103. def eval_model function for evaluating model

### DISCUSSION THREAD

**Discussion by Gunjan-Patel-16** [mrdbourke/pytorch-deep-learning#840] [Q&A]
Getting problem at accuracy
`
acc += accuracy_fn(y_true=y, y_pred = y_pred.argmax(dim=1))

model_0_results = eval_model(model=model_0,
                             data_loader=test_dataloader,
                             loss_fn=loss_fn,
                             accuracy_fn=accuracy_fn)
model_0_results`

error: 78.45
    107     """
--> 108     correct = torch.eq(y_true, y_pred).sum().item()
    109     acc = (correct / len(y_pred)) * 100
    110     return acc

RuntimeError: The size of tensor a (16) must match the size of tensor b (32) at non-singleton dimension 0

##Last 3 batch sizes 

X shape: torch.Size([32, 1, 28, 28]) | y shape: torch.Size([32]) | y_pred shape: torch.Size([32])
 X shape: torch.Size([32, 1, 28, 28]) | y shape: torch.Size([32]) | y_pred shape: torch.Size([32])
 X shape: torch.Size([16, 1, 28, 28]) | y shape: torch.Size([16]) | y_pred shape: torch.Size([16])

is last one causes the issue???

**Comment by LuluW8071** [comment]
```python 
RuntimeError: The size of tensor a (16) must match the size of tensor b (32) at non-singleton dimension 0
````

From this error 
Your `X` (probably image) has `batch_size = 16` while your `y` (image label) has `batch_size = 32` causing shape mismatch error at non-singleton dimension 0 which is index position for batch_size

**Reply by Gunjan-Patel-16** [reply]
Hello @LuluW8071 ,I appreciate your interest for helping me out with this.

From the error i also thought same way ,But the thing is shapes of data are in same manner .I've verified it by digging further in code and printing shapes of data.

Here is last 3 batches and their shapes:

X shape: torch.Size([32, 1, 28, 28]) | y shape: torch.Size([32]) | y_pred shape: torch.Size([32])
X shape: torch.Size([32, 1, 28, 28]) | y shape: torch.Size([32]) | y_pred shape: torch.Size([32])
X shape: torch.Size([16, 1, 28, 28]) | y shape: torch.Size([16]) | y_pred shape: torch.Size([16])

**Reply by LuluW8071** [reply]
The batch size is not at error from the looks of your 3 test batches. Since both X, y are returning same size.
There is may be some misplace or mistakes in ur code so I may have to see the code. 

If you are able to give the link to your code and if it reproduces the same error I may be able to help and dig out the root of causing shape error. :)

**Reply by LuluW8071** [reply]
Also just saw 
```python 
model_0_results = eval_model(model=model_0,
                        data_loader=test_dataloader,
                        loss_fn=loss_fn,
                        accuracy_fn=accuracy_fn)
model_0_results
```
which part of code is this 
Is it the training model one?

**Reply by Gunjan-Patel-16** [reply]
No its not Training model .its just basic function with evaluates model.
And here i provided link of my colab book you can go through it.

https://colab.research.google.com/drive/1VqysgzIsbBWxaz2vUudfFP2qbCYfkPEt#scrollTo=o-AeHjysDZ1v&uniqifier=2

**Reply by LuluW8071** [reply]
Unable to get access

**Reply by LuluW8071** [reply]
### Typo error:
Check in your `eval_model` code where it should have been `y_preds` instead of `y_pred` since you have defined `y_preds = model(X)`

```python
acc += accuracy_fn(y_true=y,
                         y_pred = y_pred.argmax(dim=1))
```
#### Correct way
```python
def eval_model(model:torch.nn.Module,
              data_loader:torch.utils.data.DataLoader,
              loss_fn:torch.nn.Module,
              accuracy_fn = accuracy_fn,
              device = device):
  """Returns a dictionary containing the result of model predicting on data_loader"""
  loss,acc =0,0
  model.eval()
  with torch.inference_mode():
    for X,y in data_loader:
      
      # Make prediction
      y_preds = model(X)
      # Accumulate the loss and acc values per batch
      loss += loss_fn(y_preds,y)
      # print(f" X shape: {X.shape} | y shape: {y.shape} | y_pred shape: {y_preds.argmax(dim=1).shape}")
      acc += accuracy_fn(y_true=y,
                         y_pred = y_preds.argmax(dim=1))    # <------y_preds instead of y_pred

    # Scale loss and Acc to find average loss/acc per batch
    loss /= len(data_loader)
    acc /= len(data_loader)

  return {"Model_Name ": model.__class__.__name__,# only works when model was created with class
          "Model_loss ": loss.item(),
          "Model_Acc " : acc}
```

There is also same typo error in function `eval_model_2` replace `acc += accuracy_fn(y_true=y, y_pred = y_preds.argmax(dim=1))` 
</hr>

### Also on `cell 73` of your notebook do this to avoid `device error`:
This is because the `rand_image_tensor` you passed on 'gpu' directly makes the model to call the `rand_image_tensor` from `cpu` while its on `gpu` and vice-versa. Hence the **device error**.

#### Corrected code
``` python
rand_image_tensor = torch.randn(size=(1,28,28)).to(device)
check_model(rand_image_tensor.unsqueeze(dim=0).to('cpu'))
```
#### or 
``` python
rand_image_tensor = torch.randn(size=(1,28,28))
check_model(rand_image_tensor.unsqueeze(dim=0)
```

</hr>

### Cell `97` Use `eval_model_2` function instead of `eval_model`
```python
model_2_result = eval_model_2(
    model = model_2,
    data_loader = test_dataloader,
    loss_fn = loss_fn,
    accuracy_fn = accuracy_fn,
    device = device
)
```
</hr>

### Cell `Block 112` Check you `eval_model_2` code 
You passed `Model_Name ` and `Model_Acc ` as Dictionary causing another typo error when plotting since there needs to be space in `compare_results.set_index("Model_Name ")["Model_Acc "].plot(kind = "barh")`
```python
# Visulize our model results
compare_results.set_index("Model_Name ")["Model_Acc "].plot(kind = "barh")
plt.xlabel("Accuracy (%)")
plt.ylabel("model")
```

### Close the discussion as answered if the error has been resolved :)

**Reply by Gunjan-Patel-16** [reply]
It  works perfectly, thank you for helping me out.
I'd love to connect you for more learning.
